src/guardian/keys-validation/keys-validation.service.ts

```diff
@@ -46,91 +43,92 @@ export class KeysValidationService {
 
+    this.updateCache(validatedDepositKeyList);
+
+    const invalidKeys = this.filterInvalidKeys(validatedDepositKeyList);
+
+    return [...cachedInvalidKeyList, ...invalidKeys];
+  }
+
+  filterInvalidKeys(
+    validatedKeys: [DepositKey & Key, boolean][],
+  ): RegistryKey[] {
     return validatedKeys.reduce<RegistryKey[]>(
       (invalidKeys, [data, isValid]) => {
         if (!isValid) {
-          const matchingInvalidKeys = keys.filter(
-            (key) =>
-              key.key === data.key &&
-              key.depositSignature === data.depositSignature,
-          );
-          invalidKeys.push(...matchingInvalidKeys);
+          invalidKeys.push({
+            key: data.key,
+            depositSignature: data.depositSignature,
+            operatorIndex: data.operatorIndex,
+            used: data.used,
+            index: data.index,
+            moduleAddress: data.moduleAddress,
+          });
         }
         return invalidKeys;
       },
       [],
     );
   }
 
-  /*
-   * Validate data with use of cache
-   */
-  public async validateKeys(
-    depositDataList: DepositData[],
-  ): Promise<[Key & DepositData, boolean][]> {
-    const { cachedDepositData, uncachedDepositData } =
-      this.partitionCachedData(depositDataList);
-
-    this.logger.log('Validation status of deposit keys:', {
-      cachedKeysCount: cachedDepositData.length,
-      keysNeedingValidationCount: uncachedDepositData.length,
-      totalKeysCount: depositDataList.length,
-    });
-
-    const validatedDepositData: [Key & DepositData, boolean][] =
-      await this.keyValidator.validateKeys(uncachedDepositData);
-
-    this.updateCache(validatedDepositData);
-
-    return [...cachedDepositData, ...validatedDepositData];
-  }
-
   /**
    * Partition the deposit data into cached invalid data and uncached data.
    * @param depositDataList List of deposit data to check against the cache
    * @returns An object containing cached invalid data and uncached data
    */
-  private partitionCachedData(depositDataList: DepositData[]): {
-    cachedDepositData: [DepositData, boolean][];
-    uncachedDepositData: DepositData[];
+  private partitionCachedData(
+    keys: RegistryKey[],
+    withdrawalCredentialsBuffer: WithdrawalCredentialsBuffer,
+    genesisForkVersionBuffer: Buffer,
+  ): {
+    cachedInvalidKeyList: RegistryKey[];
+    uncachedDepositKeyList: DepositKey[];
   } {
-    return depositDataList.reduce<{
-      cachedDepositData: [DepositData, boolean][];
-      uncachedDepositData: DepositData[];
+    return keys.reduce<{
+      cachedInvalidKeyList: RegistryKey[];
+      uncachedDepositKeyList: DepositKey[];
     }>(
-      (acc, depositData) => {
-        const cacheResult = this.getCachedDepositData(depositData);
-
-        if (cacheResult === false || cacheResult === true) {
-          acc.cachedDepositData.push([depositData, cacheResult]);
+      (acc, key) => {
+        const depositKey = {
```

Comment:
The main difference here is that we take as part of the search key withdrawalCredentials and genesisForkVersion?
Is there anything else to look out for when reading?

Reply:
withdrawalCredentials and genesisForkVersion as part of key we used before too
we use as key whole deposit data for caching 
But in this pr we use 
RegistryKey & {
  withdrawalCredentials: WithdrawalCredentialsBuffer;
  genesisForkVersion: Buffer;
}; type for validation

to get back list of RegistryKey & {
  withdrawalCredentials: WithdrawalCredentialsBuffer;
  genesisForkVersion: Buffer;
};

and just map result to RegistryKey

Reply:
prev we needed to make a search across result of validation to find all matching RegistryKey